File: benchmarks/split.go
package main

import (
	"bytes"
	"fmt"
	"os"
)

func isSpace(b byte) bool {
	return b == ' ' || (b >= 9 && b <= 13)
}

func main() {
	mode := "byte"
	if len(os.Args) > 1 {
		mode = os.Args[1]
	}
	const N = 1_000_000_000
	const NEEDLE byte = 7

	data := make([]byte, N)
	for i := 0; i < N; i++ {
		data[i] = byte(i & 0xff)
	}

	var seps int
	switch mode {
	case "byte":
		// bytes.Count dispatches to hand-written SIMD (NEON on arm64,
		// AVX2/SSE on amd64); that's the idiomatic fast primitive.
		seps = bytes.Count(data, []byte{NEEDLE})
	case "predicate":
		// No stdlib equivalent for a custom predicate, so scalar loop.
		for _, b := range data {
			if isSpace(b) {
				seps++
			}
		}
	default:
		fmt.Fprintf(os.Stderr, "unknown mode: %s\n", mode)
		os.Exit(1)
	}
	fmt.Println(seps + 1)
}

File: benchmarks/split.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define BLACK_BOX(x) __asm__ volatile("" : "+r" (x))

static inline uint64_t hash(uint64_t x) {
    // This forces clang to not optimize the first part of the calculations away.
    // It might feel like cheating by handcuffing C but we don't want to test how
    // good clang can optimize the test-data.
    BLACK_BOX(x);
    x ^= x >> 30;
    x *= 0xbf58476d1ce4e5b9ULL;
    x ^= x >> 27;
    x *= 0x94d049bb133111ebULL;
    x ^= x >> 31;
    return x;
}

static inline int is_space(uint8_t b) {
    return b == ' ' || (b >= 9 && b <= 13);
}

int main(int argc, char **argv) {
    const char *mode = argc > 1 ? argv[1] : "byte";
    const size_t N = 500000000;
    const uint8_t NEEDLE = 7;

    uint8_t *data = (uint8_t *)malloc(N);
    if (!data) return 1;
    for (size_t i = 0; i < N; i++) {
        data[i] = (uint8_t)(hash((uint64_t)i) & 0xff);
    }

    // Number of split pieces = number of separators + 1.
    // The branchless form lets clang autovectorize with NEON.
    size_t seps = 0;
    if (!strcmp(mode, "byte")) {
        for (size_t i = 0; i < N; i++) {
            seps += (data[i] == NEEDLE);
        }
    } else if (!strcmp(mode, "predicate")) {
        for (size_t i = 0; i < N; i++) {
            seps += is_space(data[i]);
        }
    } else {
        fprintf(stderr, "unknown mode: %s\n", mode);
        free(data);
        return 1;
    }
    printf("%zu\n", seps + 1);

    free(data);
    return 0;
}
